src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    internal sealed class RemoteJSDataStream : Stream
+    {
+        private readonly RemoteJSRuntime _runtime;
+        private readonly long _streamId;
+        private readonly long _totalLength;
+        private readonly CancellationToken _streamCancellationToken;
+        private readonly Stream _pipeReaderStream;
+        private readonly Pipe _pipe;
+        private long _bytesRead;
+
+        public static async Task<bool> ReceiveData(RemoteJSRuntime runtime, long streamId, byte[] chunk, string error)
+        {
+            if (!runtime.RemoteJSDataStreamInstances.TryGetValue(streamId, out var instance))
+            {
+                // There is no data stream with the given identifier. It may have already been disposed.
+                // We notify JS that the stream has been cancelled/disposed.
+                return false;
+            }
+
+            return await instance.ReceiveData(chunk, error);
+        }
+
+        public static async ValueTask<RemoteJSDataStream> CreateRemoteJSDataStreamAsync(
+            RemoteJSRuntime runtime,
+            IJSDataReference jsDataReference,
+            long totalLength,
+            long maxBufferSize,
+            long maximumIncomingBytes,
+            CancellationToken cancellationToken = default)
+        {
+            var streamId = runtime.RemoteJSDataStreamNextInstanceId++;
+            var remoteJSDataStream = new RemoteJSDataStream(runtime, streamId, totalLength, maxBufferSize, cancellationToken);
+            await runtime.InvokeVoidAsync("Blazor._internal.sendJSDataStream", jsDataReference, streamId, maximumIncomingBytes);
```

Comment:
We might want to use this call to send some data to the client like the max message size (if it doesn't have it) and the maximum number of chunks to send ahead of time so that those values don't have to be hard-coded on the JS side.

Reply:
The pipe likely takes care of it, but we might want to think about "fragmentation" and determine if its convenient for us to enforce a "chunk" size on the client that cleanly maps to "slabs" on the server memory we are renting.

Reply:
> max message size

That's `maximumIncomingBytes` which is the SignalR message size limit (32kb).

> and the maximum number of chunks to send ahead of time

By "ahead of time" I'm interpreting that as between acks? If so, I think the [technique we use currently](https://github.com/dotnet/aspnetcore/pull/33491/files/62f2338e1a8bf907ecae707bc604da30593c6d33#diff-7c47c684c78e3f5b5eeecb28b249b03afcefcd90a925cc20b69ec2cdec261752R166-R171) (dynamically setting num chunks before next ack based on speed/interactivity may provide a better user experience).

Reply:
> enforce a "chunk" size on the client that cleanly maps to "slabs" on the server memory we are renting

Based on how we currently set the chunk size, it'll be the same for all calls (except the final chunk which may be smaller). Hence we should be able to re-use rented "slabs" (though we may have to implement the re-use logic into Blazor MessagePack).

Reply:
> The pipe likely takes care of it, but we might want to think about "fragmentation" and determine if its convenient for us to enforce a "chunk" size on the client that cleanly maps to "slabs" on the server memory we are renting.

I thought `Pipe` abstracted us from those details, and made sensible choices about buffers etc. If we can avoid reinventing any of those low-level details (i.e., as long as we're not going to create any severe perf problems) that would be ideal. We don't need to be 100% optimal - better to let `Pipe` evolve to improve over time.